### PR TITLE
feat: add common distance calculation

### PR DESCRIPTION
**Pull request by noneback**
### Purpose

Describe the problem solved or feature added by this PR. In addition, please provide a link to the issue this PR is related to, if applicable.

### Approach

How does this PR solve the problem or add the feature? What are the major changes in the codebase introduced by this PR?

### Testing

- [x] I have tested this PR locally.
- [x] I added tests to cover my changes, if not applicable, I have added a reason why.

How did you test this PR? Please provide a description of the tests that you ran to verify your changes.

How should the reviewer test this PR? If applicable, have you added tests to cover your changes?

### Chore checklist

- [x] I have updated the documentation accordingly.
- [x] I have added comments to most of my code, particularly in hard-to-understand areas.

### CODE REVIEW

File: src/func/collection.rs [hunk 1]
```diff
@@ -22,8 +27,20 @@ pub struct Config {
...
     #[new]
-    pub fn new(ef_construction: usize, ef_search: usize, ml: f32) -> Self {
-        Self { ef_construction, ef_search, ml }
+    pub fn new(
+        ef_construction: usize,
+        ef_search: usize,
+        ml: f32,
+        distance: &str,
+    ) -> Result<Self, Error> {
+        Distance::from(&distance.into())?;
```
Comment (edwinkys) [inline src/func/collection.rs:36]: I assume that this line is for validation?

File: src/func/collection.rs [hunk 2]
```diff
@@ -22,8 +27,20 @@ pub struct Config {
...
+
+        Ok(Self {
+            ef_construction: ef_construction,
+            ef_search: ef_search,
+            ml: ml,
```
Comment (edwinkys) [inline src/func/collection.rs:41]: According to clippy, you need to use the shorthand version like just `ef_construction`.

File: src/func/collection.rs [hunk 3]
```diff
@@ -619,3 +636,23 @@ impl SearchResult {
...
+    }
+
+    #[test]
+    #[should_panic(expected = "Distance not supported")]
+    fn config_with_distance_panic() {
+        for dist in vec!["l2", "test"] {
+            let conf = Config::new(10, 10, 3.0, dist).unwrap();
+        }
+    }
+}
```
Comment (edwinkys) [inline src/func/collection.rs:658]: Can we move the test to `tests/test_collection`? We want to keep this file clean and dedicated for the functionality.

File: src/func/distance.rs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+            _ => Err("Distance not supported".into()),
+        }
+    }
+
+    pub fn calculate(&self, a: &Vector, b: &Vector) -> f32 {
+        assert_eq!(a.0.len(), b.0.len());
+
+        match self {
+            Distance::Dot => {
+                a.0.iter().zip(b.0.iter()).map(|(x, y)| x * y).sum::<f32>()
```
Comment (edwinkys) [inline src/func/distance.rs:27]: Can we have this in separate function like `calculate_dot`?

File: src/func/distance.rs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            }
+            Distance::Euclidean => {
+                a.0.iter()
+                    .zip(b.0.iter())
+                    .map(|(a, b)| (a - b).powi(2))
+                    .sum::<f32>()
+                    .sqrt()
```
Comment (edwinkys) [inline src/func/distance.rs:34]: Just like the dot calculation, can we have this in a separate function?

File: src/func/distance.rs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            }
+            Distance::Cosine => {
+                let dot_product: f32 =
+                    a.0.iter().zip(b.0.iter()).map(|(x, y)| x * y).sum();
+                let magnitude_a: f32 =
+                    a.0.iter().map(|x| x.powi(2)).sum::<f32>().sqrt();
+                let magnitude_b: f32 =
+                    b.0.iter().map(|y| y.powi(2)).sum::<f32>().sqrt();
+
+                dot_product / (magnitude_a * magnitude_b)
```
Comment (edwinkys) [inline src/func/distance.rs:44]: Can we have this in separate function like `calculate_cosine`?

**Review by edwinkys — CHANGES_REQUESTED** [review]
All is looking great! I just have some things that I requested and also, double check the PR with `cargo clippy` to make sure it passes the checks.

File: src/func/collection.rs (hunk 1 above)
Comment (noneback) [reply src/func/collection.rs:36]: yes, we valid it when conf is created.

File: src/func/collection.rs (hunk 2 above)
Comment (noneback) [reply src/func/collection.rs:41]: I will do cargo clippy to check those code.

File: src/func/collection.rs (hunk 3 above)
Comment (noneback) [reply src/func/collection.rs:658]: Sure, I will fix that.

File: src/func/distance.rs (hunk 6 above)
Comment (noneback) [reply src/func/distance.rs:44]: Sure

**Comment by edwinkys** [comment]
The Python tests seems to be failing since we add the new distance configuration.

**Comment by noneback** [comment]
> The Python tests seems to be failing since we add the new distance configuration.

I pushed some codes to fix.

**Review by edwinkys — APPROVED** [review]
Everything is looking great! I'll merge it tomorrow morning here. It's quite late at night here 😁

**Pull request merged**